doc/man/gnunet-peer-info.1
.TH GNUNET-PEER-INFO 1 "May 6, 2008" "GNUnet"

.SH NAME
gnunet\-peer\-info \- Display information about other known nodes.

.SH SYNOPSIS
.B gnunet\-peer\-info
.RI [ options ]
.br

.SH DESCRIPTION
.PP
\fBgnunet\-peer\-info\fP display the last known IP and GNUnet host address of known nodes.

.SH OPTIONS
.B
.IP "\-c FILENAME, \-\-config=FILENAME"
Load config file (default: /etc/gnunet.conf)
.B
.IP "\-h, \-\-help"
Print help page
.B
.IP "\-H HOSTNAME, \-\-host=HOSTNAME"
Specify host on which gnunetd is running
.B
.IP "\-n, \-\-numeric"
Disable resolution of IPs to hostnames
.B
.IP "\-q, \-\-quiet"
Do not print anything but the peer identities
.B
.IP "\-s, \-\-self"
Print only our own identity (together with "\-q", this is the exact line that other peers would have to put in to their friends file in order to consider this peer one of their friends in F2F mode).
.B
.IP "\-v, \-\-version"
Print the version number
.B
.IP "\-L LOGLEVEL, \-\-loglelvel=LOGLEVEL"
Set the loglevel

.SH BUGS
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>

.SH SEE ALSO
gnunetd.conf(5), gnunetd(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/server/gnunet-peer-info.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2003, 2004, 2006 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file server/gnunet-peer-info.c
 * @brief Print information about other known peers.
 * @author Christian Grothoff
 */

#include "platform.h"
#include "gnunet_util.h"
#include "gnunet_protocols.h"
#include "gnunet_directories.h"
#include "gnunet_transport_service.h"
#include "gnunet_identity_service.h"
#include "gnunet_core.h"
#include "core.h"

static GNUNET_Transport_ServiceAPI *transport;

static GNUNET_Identity_ServiceAPI *identity;

static struct GNUNET_GE_Context *ectx;

static char *cfgFilename = GNUNET_DEFAULT_DAEMON_CONFIG_FILE;

static int no_resolve = GNUNET_NO;

static int get_self = GNUNET_NO;

static int be_quiet = GNUNET_NO;

/**
 * All gnunet-peer-info command line options
 */
static struct GNUNET_CommandLineOption gnunetpeerinfoOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Print information about GNUnet peers.")),     /* -h */
  GNUNET_COMMAND_LINE_OPTION_HOSTNAME,  /* -H */
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'n', "numeric", NULL,
   gettext_noop ("don't resolve host names"),
   0, &GNUNET_getopt_configure_set_one, &no_resolve},
  {'q', "quiet", NULL,
   gettext_noop ("output only the identity strings"),
   0, &GNUNET_getopt_configure_set_one, &be_quiet},
  {'s', "self", NULL,
   gettext_noop ("output our own identity only"),
   0, &GNUNET_getopt_configure_set_one, &get_self},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_END,
};

#if HAVE_ADNS | HAVE_C_ARES
/**
 * Prepass just to resolve DNS entries.
 */
static int
resolveHostInfo (const GNUNET_PeerIdentity * id,
                 const unsigned short proto, int verified, void *data)
{
  GNUNET_MessageHello *hello;
  void *addr;
  unsigned int addr_len;
  char *info;
  int have_addr;

  if (GNUNET_shutdown_test () == GNUNET_YES)
    return GNUNET_SYSERR;
  hello = identity->identity2Hello (id, proto, GNUNET_NO);
  if (NULL == hello)
    return GNUNET_OK;
  addr = NULL;
  addr_len = 0;
  have_addr = transport->hello_to_address (hello, &addr, &addr_len);
  GNUNET_free (hello);
  if (have_addr == GNUNET_OK)
    {
      info = GNUNET_get_ip_as_string (addr, addr_len, !no_resolve);
      GNUNET_free (addr);
      addr = NULL;
      GNUNET_free_non_null (info);
    }
  return GNUNET_OK;
}

#endif

/**
 * Print information about the peer.
 * Currently prints the GNUNET_PeerIdentity, trust and the IP.
 * Could of course do more (e.g. resolve via DNS).
 */
static int
printHostInfo (const GNUNET_PeerIdentity * id,
               const unsigned short proto, int verified, void *data)
{
  GNUNET_MessageHello *hello;
  void *addr;
  unsigned int addr_len;
  char *info;
  int have_addr;
  GNUNET_EncName enc;

  if (GNUNET_shutdown_test () == GNUNET_YES)
    return GNUNET_SYSERR;
  GNUNET_hash_to_enc (&id->hashPubKey, &enc);
  hello = identity->identity2Hello (id, proto, GNUNET_NO);
  if (NULL == hello)
    {
      GNUNET_GE_LOG (ectx,
                     GNUNET_GE_WARNING | GNUNET_GE_BULK | GNUNET_GE_USER,
                     _("Could not get address of peer `%s'.\n"), &enc);
      return GNUNET_OK;
    }
  if (GNUNET_SYSERR == GNUNET_RSA_verify (&hello->senderIdentity,
                                          GNUNET_sizeof_hello (hello) -
                                          sizeof (GNUNET_RSA_Signature) -
                                          sizeof (GNUNET_RSA_PublicKey) -
                                          sizeof (GNUNET_MessageHeader),
                                          &hello->signature,
                                          &hello->publicKey))
    {
      GNUNET_GE_LOG (ectx,
                     GNUNET_GE_WARNING | GNUNET_GE_BULK | GNUNET_GE_USER,
                     _("`%s' message invalid (signature invalid).\n"),
                     "HELLO");
    }
  addr = NULL;
  addr_len = 0;
  have_addr = transport->hello_to_address (hello, &addr, &addr_len);
  GNUNET_free (hello);
  if (have_addr != GNUNET_OK)
    {
      info = GNUNET_strdup ("NAT");     /* most likely */
    }
  else
    {
      info = GNUNET_get_ip_as_string (addr, addr_len, !no_resolve);
      GNUNET_free (addr);
      addr = NULL;
    }
  if (info == NULL)
    {
      GNUNET_GE_LOG (ectx,
                     GNUNET_GE_DEBUG | GNUNET_GE_BULK | GNUNET_GE_USER,
                     _("Could not get address of peer `%s'.\n"), &enc);
      if (be_quiet)
        printf ("%s\n", (char *) &enc);
      else
        printf (_("Peer `%s' with trust %8u\n"),
                (char *) &enc, identity->getHostTrust (id));
      return GNUNET_OK;
    }
  if (be_quiet)
    printf ("%s\n", (char *) &enc);
  else
    printf (_("Peer `%s' with trust %8u and address `%s'\n"),
            (char *) &enc, identity->getHostTrust (id), info);
  GNUNET_free (info);
  return GNUNET_OK;
}

int
main (int argc, char *const *argv)
{
  struct GNUNET_GC_Configuration *cfg;
  struct GNUNET_CronManager *cron;
  const GNUNET_RSA_PublicKey *me;
  GNUNET_PeerIdentity id;
  GNUNET_EncName enc;
  int ret;

  ret = GNUNET_init (argc,
                     argv,
                     "gnunet-peer-info",
                     &cfgFilename, gnunetpeerinfoOptions, &ectx, &cfg);
  if (ret == -1)
    {
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  GNUNET_GE_ASSERT (ectx,
                    0 == GNUNET_GC_set_configuration_value_string (cfg,
                                                                   ectx,
                                                                   "TCPSERVER",
                                                                   "DISABLE",
                                                                   "YES"));
  cron = GNUNET_cron_create (ectx);
  if (GNUNET_OK != GNUNET_CORE_init (ectx, cfg, cron, NULL))
    {
      GNUNET_cron_destroy (cron);
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  identity = GNUNET_CORE_request_service ("identity");
  if (identity == NULL)
    {
      GNUNET_CORE_done ();
      GNUNET_cron_destroy (cron);
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  transport = GNUNET_CORE_request_service ("transport");
  if (transport == NULL)
    {
      GNUNET_CORE_release_service (identity);
      GNUNET_CORE_done ();
      GNUNET_cron_destroy (cron);
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  if (get_self != GNUNET_YES)
    {
      if (no_resolve != GNUNET_YES)
        {
#if HAVE_ADNS | HAVE_C_ARES
          identity->forEachHost (0,     /* no timeout */
                                 &resolveHostInfo, NULL);
          /* give GNU ADNS time to resolve... */
          GNUNET_thread_sleep (2 * GNUNET_CRON_SECONDS);
#endif
        }
      identity->forEachHost (0, /* no timeout */
                             &printHostInfo, NULL);
    }
  else
    {
      me = identity->getPublicPrivateKey ();
      identity->getPeerIdentity (me, &id);
      GNUNET_hash_to_enc (&id.hashPubKey, &enc);
      if (be_quiet)
        printf ("%s\n", (char *) &enc);
      else
        printf (_("I am peer `%s'.\n"), (const char *) &enc);
    }
  GNUNET_CORE_release_service (identity);
  GNUNET_CORE_release_service (transport);
  GNUNET_CORE_done ();
  GNUNET_cron_destroy (cron);
  GNUNET_fini (ectx, cfg);
  return 0;
}


/* end of gnunet-peer-info.c */
